### ISSUE TITLE
Redirect in ActionDispatch::Integration::Session

### ISSUE THREAD

**Issue by kingdonb** [biola/rack-cas#77]
Hello, I recently learned about this ActionDispatch::Integration::Session which is the object that you find yourself inside of (I guess?) when you are running an Integration test in `Test::Unit`.  From the `rails console` executed as a standalone process you can access this object as 'app' – neat!

So I wanted to try and use this with my app, to do some interactive debugging in a way without starting up a full server.  The first hurdle I found of course is that CAS wants me to log in unless I disable it... I thought there might be a way to manipulate the session interactively so that I wouldn't need to disable it, but I found nothing like that.

What I wound up doing is enable FakeCAS, to get around the need to use an actual CAS server, but still putting the right things in my session so that helpers and controllers that expect to have a proper user session set up before they run, don't choke.  I figured I can send a POST to '/login' like the FakeCAS login prompt does....

I tried having a conversation to establish a session like this, and this is what the result is like:

```
2.3.0 :001 > app.get '/'


Started GET "/" for 127.0.0.1 at 2017-07-18 15:48:08 -0400
  ActiveRecord::SchemaMigration Load (30.6ms)  SELECT "schema_migrations".* FROM "schema_migrations"
Processing by IndexController#index as HTML
  Rendered text template (0.0ms)
Filter chain halted as :authenticate rendered or redirected
Completed 401 Unauthorized in 23ms (Views: 11.9ms | ActiveRecord: 0.0ms)
 => 200
2.3.0 :002 > puts app.response.body
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Fake CAS</title>
  </head>
  <body>
    <form action="/login" method="post">
      <input type="hidden" name="service" value="http://www.example.com/"/>
      <label for="username">Username</label>
      <input id="username" name="username" type="text"/>
      <label for="password">Password</label>
      <input id="password" name="password" type="password"/>
      <input type="submit" value="Login"/>
    </form>
  </body>
</html>
 => nil
2.3.0 :003 > app.post '/login', params: {username: 'KINGDON', password: 'aosidfj', service: 'http://localhost:3000'}


Started POST "/login" for 127.0.0.1 at 2017-07-18 15:48:37 -0400
 => 302
2.3.0 :004 > app.response.redirect_url
 => nil
2.3.0 :005 > app.response.body
 => "Redirecting you..."
2.3.0 :006 > app.session['cas']
 => {"user"=>nil, "extra_attributes"=>{}}
```

I'm sure I'm not exactly meant to use rack-cas in this way, but is this a bug?

Look at that redirect in particular.  I looked at the response.headers and it looks like `Location` is unset.  What kind of redirect is that?

I'm asking if it's a bug (rather than telling) because when I do this exchange with a browser, I get redirected properly (and I don't know why, if you can answer that one thing for me it might be all I need to know...) When I do the exchange with a rails console, I just seem to be missing something.  I tried setting app.session['cas']={'user' => 'KINGDON'} here but it does not seem to be effective, I think it is just discarded information when the next request is handled.

I could set the cookie I think, but it's encrypted and that seems even more cumbersome.

So, the next thing I tried was adding a pry statement to the `user_login` method, the second of the `before_action`s that I call after `authenticate` in my ApplicationController.  This is the method that makes sure I got back a valid user who is actually authorized and setup their session.

```
2.3.0 :007 > app.get '/'


Started GET "/" for 127.0.0.1 at 2017-07-18 15:51:30 -0400
Processing by IndexController#index as HTML

From: /Users/kingdonb/Desktop/devel/pa/pa_phase2/app/controllers/application_controller.rb @ line 19 ApplicationController#user_login:

    18: def user_login
 => 19:   binding.pry
    20:   if session['cas']['user'].nil?
    21:     login_err("Your single sign on authentication could not be verified.")
    22:     return false
...

> session['cas']['user'] = 'KINGDON'
> ^D
...

> app.get '/'


Started GET "/" for 127.0.0.1 at 2017-07-18 15:51:30 -0400
Processing by IndexController#index as HTML

From: /Users/kingdonb/Desktop/devel/pa/pa_phase2/app/controllers/application_controller.rb @ line 19 ApplicationController#user_login:

    18: def user_login
 => 19:   binding.pry
...
> ^D

2.3.0 :014 > app.session['cas']
 => {"user"=>"KINGDON", "extra_attributes"=>{}}
```

That looks like it worked, now I can have an interactive session and issue all of the `app.get` calls I want, the session will be persisted.  It just seemed like there should be an easier way to get there.

Do you have any ideas for me about how to handle this better?  Is there a better way to bootstrap a request and get a response context from inside of Rails Console?  I apologize that this issue is a little outside of what's probably a supported use of `rack-cas`, but I was hoping maybe you could answer some of that, or at least help me understand how I'm getting a session with `user` set to nil?

**Comment by kingdonb** [comment]
Interesting, this works better:

```
2.3.0 :001 > app.post '/login?username=MSCOTT4&password=aosidfj&service=http://localhost:3000'
```

It looks like `post '/path', params: {params: 'hash'}` is an old way of calling post, with params that just go into params_hash now... I just needed to try this instead:

```ruby
> app.post '/login', {params: 'hash'}
```

**Comment by kingdonb** [comment]
I solved my own problem.

So, I think I needed to invoke FakeCAS to do this.  I'd be interested in a way that I could do this without hitting FakeCAS (I'm not usually setting up FakeCAS in my development environment, so I'd be interested to know if there is a way)

But for my issue, I can close it.  Simple syntax mistake, changing from params: {hash} to {params: hash} solved everything for me.  Thanks for reading along.
